### PR TITLE
Use Vector in trie traversal

### LINKED ISSUE #3098: Testing of LFS
## Overview

This issue is to record several rounds of testing of **Last Finalized State** pre release versions. Tests are more performance oriented and some of the bug fixes and optimitations are done between each test setup.

Results are starting from the last setup. Shows how speed is increased with stronger machine. This is due to additonal computation needed for verification of the Rholang state (#3145). This is visible in high usage of CPU during sync.  

Another important factor is disk speed. All tests are done on SSD disks and on stronger machines faster speed reduces overall sync duration.

### Testing setup 4

This round is almost identical to _setup 3_ with additional optimization in receiving blocks (#3243) which include block sorting as part of download and with testing on 32 vCPU machine.

RNode Docker image version: [`rchain/rnode:v0.9.26-rc`](https://hub.docker.com/layers/rchain/rnode/v0.9.26-rc/images/sha256-af9d5c6467474eb6e47d1557dfdef26c312a5834eeb3fed5c94b8d6ce25a5ee9?context=explore)

| Nr. | Sync time<br>(fully catch up) | Sync Rholang state | Sync blocks | VPS | Log
| --- | ---        | ---    | ---    | ---                         | ---
| 1   | **2.25 h** | 1.5 h  | 1.75 h | 32vCPU 64GB (DigitalOcean)  | [test-4-do-32vcpu-64g.zip](https://github.com/rchain/rchain/files/5630012/test-4-do-32vcpu-64g.zip)
| 2   | **3.25 h** | 1.5 h  | 2 h    | 16vCPU 32GB CPX51 (Hetzner) | [test-4-cpx51-32g.zip](https://github.com/rchain/rchain/files/5630016/test-4-cpx51-32g.zip)
| 3   | **3 h**    | 1.75 h | 2 h    | 8vCPU 32GB (IBM)                   | [test-4-ibm-8vcpu-32g.zip](https://github.com/rchain/rchain/files/5630025/test-4-ibm-8vcpu-32g.zip)
| 4   | **4 h**    | 2.25 h | 2.5 h  | 8vCPU 16GB CPX41 (Hetzner)  | [test-4-cpx41-16g.zip](https://github.com/rchain/rchain/files/5630022/test-4-cpx41-16g.zip)
| 5   | **8 h**    | 4.5 h  | 3.5 h  | 4vCPU 8GB CPX31 (Hetzner)   | [test-4-cpx31-8g.zip](https://github.com/rchain/rchain/files/5630030/test-4-cpx31-8g.zip)

NOTE: CPU and memory observed in testing is the same as in setup 3, only overall duration is shorter.

[1] CPU, memory - 32vCPU 64GB (DigitalOcean)
test-4-do-32vcpu-64g-lfs-progress-cpu

### Testing setup 3

First two rounds of testing showed that LFS source node disk reading speed can make significant difference in overall speed for nodes to download Rholang state. In this testing setup,  to make reading from disk much faster RSpace folder on source node is converted to RAM disk which reduced reading from ~20sec to ~1sec for chunk of state sent over the network.

RNode Docker image version: [`tgrospic/rnode:v0.9.26-rc`](https://hub.docker.com/layers/tgrospic/rnode/v0.9.26-rc/images/sha256-85bb8d444b354b0628d871b561c2e908131f79cef880ece5b86678f64fb49195?context=repo)

NOTE: Bug with memory leak on network errors is still not resolved and one test with 8GB machine failed not listed in results.

| Nr. | Sync time<br>(fully catch up) | VPS
| --- | ---  | --- | 
| 1 | 4 hours (2.5 h blocks)<br>[test-3-cpx51-32g-full-ok-3.zip](https://github.com/rchain/rchain/files/5629935/test-3-cpx51-32g-full-ok-3.zip) | 16vCPU 32GB CPX51
| 2 | 5.5 hours (3.5 h blocks)<br>[test-3-cpx41-16g-full-ok-1.zip](https://github.com/rchain/rchain/files/5629947/test-3-cpx41-16g-full-ok-1.zip) | 8vCPU 16GB CPX41
| 3 | 9 hours (4.5 h blocks)<br>[test-3-cpx31-8g-full-ok-1.zip](https://github.com/rchain/rchain/files/5629957/test-3-cpx31-8g-full-ok-1.zip) | 4vCPU 8GB CPX31

#### Observations from testing

[1] CPU, memory - 16vCPU 32GB CPX51
test-3-cpx51-32g-lfs-progress-cpu

[1] Direct buffer - 16vCPU 32GB CPX51
test-3-cpx51-32g-lfs-progress-buffer

[2] CPU, memory - 8vCPU 16GB CPX41
test-3-16g-lfs-progress-cpu

[3] CPU, memory - 4vCPU 8GB CPX31
test-3-8g-lfs-progress-cpu

Survived network errors without crashing - 4vCPU 8GB CPX31
test-3-8g-lfs-progress-3-errors-cpu

### Testing setup 2

RNode Docker image version: [`tgrospic/rnode:v0.9.26-beta`](https://hub.docker.com/layers/tgrospic/rnode/v0.9.26-beta/images/sha256-0734a74f064939d43208d82c6914f64a0917b4de95cca0ae863a3999f2d53e3c?context=repo)

| Nr. | Source | Duration | Sync time<br>(fully catch up) | Environment
| --- | ---  | ---   | --- | ---
| 1 | Full  | 12.5 hours (2.5 h blocks) | - | One machine + 3 active nodes
| 2 | LFS  | ERROR interrupted<br>[rnode-restart-error.zip](https://github.com/rchain/rchain/files/5241197/rnode-restart-error.zip) | - | Target alone on another machine<br>(Docker on Hetzner cloud CPX31)
| 3 | LFS [2] | 4.5 hours (2.5 blocks)<br> [rnode-test-3.zip](https://github.com/rchain/rchain/files/5244082/rnode-test-3.zip) | 8 hours | Target alone on another machine<br>(Docker on Hetzner cloud CPX31)
| 4 | LFS [3] | 6 hours (2 blocks)<br> [rnode-test-4.zip](https://github.com/rchain/rchain/files/5248452/rnode-test-4.zip) | 10 hours | Target alone on another machine<br>(Docker on Hetzner cloud CX31)

#### Observations from testing

[3] After node received LFS, it's not responding to requests for tuple space (_StateItems_). After restart it works as expected.

[2] CPU, memory
image

[2] Direct buffer memory
image

### Testing setup 1

~Trie traversal is slow when number of records is bigger then 3,000 and it's rising faster then linear.~ Resolved in PR #3099.

Performance of LFS should be tested with this configuration.

RNode Docker image for all nodes: [`rchain/rnode:v0.9.26-alpha`](https://hub.docker.com/layers/rchain/rnode/v0.9.26-alpha/images/sha256-8dee109eb6bb377a8912111e09489ea105f926eb36d3d6a53e5195738d170d60?context=explore)

1. Observer node with full state
2. Observer node with empty state (bootstrap from `1.`)
3. Observer node with empty state (bootstrap from `2.`)

~Observer nodes with empty state (2. and 3.) must have higher limit for direct buffer memory to 3GB.
`-XX:MaxDirectMemorySize=3g`~
Memory leak with direct buffer is fixed in transport layer #3239 so now limit can be much lower.
`-XX:MaxDirectMemorySize=200m`

First operation is to test syncing of LFS from the full node. When this is done second syncing should use this node with _trimmed_ state as a source to sync the third node.

| Operation | Expected    | Measured
| ---       | ---         | ---
| 1. -> 2.  | 4-6 hours   | 7.5 hours (one machine + 1 active nodes)<br>11 hours (one machine + 4 active nodes)
| 2. -> 3.  | 1-3 hours   | 4 hours (one machine + 4 active nodes)

### PR DESCRIPTION
**Pull request by tgrospic**
## Overview
This PR fixes trie traversal to use Vector instead of Seq to speed up concatenation which was reason for slowdown with larger number of traversed records.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: rspace/src/main/scala/coop/rchain/rspace/state/exporters/RSpaceExporterItems.scala [hunk 1]
```diff
@@ -62,4 +55,32 @@ object RSpaceExporterItems {
...
+  def getHistoryAndData[F[_]: Sync, Value](
+      exporter: RSpaceExporter[F],
+      startPath: Seq[(Blake2b256Hash, Option[Byte])],
+      skip: Int,
+      take: Int,
+      fromBuffer: ByteBuffer => Value
+  ): F[(StoreItems[Blake2b256Hash, Value], StoreItems[Blake2b256Hash, Value])] =
+    for {
+      // Traverse and collect tuple space nodes
+      nodes <- time("TRAVERSE TRIE")(exporter.getNodes(startPath, skip, take))
```
Comment (nzpr) [inline rspace/src/main/scala/coop/rchain/rspace/state/exporters/RSpaceExporterItems.scala:68]: Just out of curiosity for now - is using this `time` in production expected? Or this should be replaced at some point with Jaeger spans?

File: casper/src/main/scala/coop/rchain/casper/engine/LastFinalizedStateTupleSpaceRequester.scala [hunk 2]
```diff
@@ -27,8 +28,9 @@ object LastFinalizedStateTupleSpaceRequester {
 
   type StatePartPath = Seq[(Blake2b256Hash, Option[Byte])]
 
-  // TODO: extract this as a parameter (3000 on local machine is showing best results)
-  val pageSize = 3000
+  // TODO: extract this as a parameter
+  // 20,000 records uses about 2G of direct buffer memory
```
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/engine/LastFinalizedStateTupleSpaceRequester.scala]: 2G is on importing node right? Not on exporting?

File: rspace/src/main/scala/coop/rchain/rspace/state/exporters/RSpaceExporterItems.scala (hunk 1 above)
Comment (tgrospic) [reply rspace/src/main/scala/coop/rchain/rspace/state/exporters/RSpaceExporterItems.scala:68]: It's temporary, should be removed before merge to dev.

File: casper/src/main/scala/coop/rchain/casper/engine/LastFinalizedStateTupleSpaceRequester.scala (hunk 2 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/engine/LastFinalizedStateTupleSpaceRequester.scala]: I've updated the comment to refer to importing side.

**Review by nzpr — APPROVED** [review]
LGTM

**Review by zsluedem — APPROVED** [review]
LGTM

**Pull request merged**